plugins/rest/rest_auth.go

```diff
@@ -113,56 +121,169 @@ type tokenEndpointResponse struct {
+		// Use client_credentials as default to not break existing config
+		ap.GrantType = "client_credentials"
+	} else if ap.GrantType != "client_credentials" && ap.GrantType != "jwt_bearer" {
+		return nil, errors.New("grant_type must be either client_credentials or jwt_bearer")
+	}
+
+	if ap.GrantType == "jwt_bearer" || (ap.GrantType == "client_credentials" && ap.SigningKeyID != "") {
+		if err = ap.parseSigningKey(c); err != nil {
+			return nil, err
+		}
+	}
+
 	// Inherit skip verify from the "parent" settings. Should this be configurable on the credentials too?
 	ap.tlsSkipVerify = c.AllowInsureTLS
 
 	if !strings.HasPrefix(ap.TokenURL, "https://") {
 		return nil, errors.New("token_url required to use https scheme")
 	}
-	if ap.ClientID == "" || ap.ClientSecret == "" {
-		return nil, errors.New("client_id and client_secret required")
-	}
-	if ap.Scopes == nil {
-		ap.Scopes = &[]string{}
+	if ap.GrantType == "client_credentials" {
+		if ap.ClientSecret != "" && ap.SigningKeyID != "" {
+			return nil, errors.New("can only use one of client_secret and signing_key for client_credentials")
+		}
+		if ap.SigningKeyID == "" && (ap.ClientID == "" || ap.ClientSecret == "") {
+			return nil, errors.New("client_id and client_secret required")
+		}
 	}
 
 	return DefaultRoundTripperClient(t, *c.ResponseHeaderTimeoutSeconds), nil
 }
 
-// requestToken tries to obtain an access token using the client credentials flow
+// requestToken tries to obtain an access token using either the client credentials flow
 // https://tools.ietf.org/html/rfc6749#section-4.4
+// or the JWT authorization grant
+// https://tools.ietf.org/html/rfc7523
 func (ap *oauth2ClientCredentialsAuthPlugin) requestToken() (*oauth2Token, error) {
-	body := url.Values{"grant_type": []string{"client_credentials"}}
-	if len(*ap.Scopes) > 0 {
-		body["scope"] = []string{strings.Join(*ap.Scopes, " ")}
+	body := url.Values{}
+	if ap.GrantType == "jwt_bearer" {
+		authJwt, err := ap.createAuthJWT(ap.Claims, ap.signingKeyParsed)
+		if err != nil {
+			return nil, err
+		}
+		body.Add("grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer")
+		body.Add("assertion", *authJwt)
+	} else {
+		body.Add("grant_type", "client_credentials")
+
+		if ap.SigningKeyID != "" {
+			authJwt, err := ap.createAuthJWT(ap.Claims, ap.signingKeyParsed)
+			if err != nil {
+				return nil, err
+			}
+			body.Add("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
+			body.Add("client_assertion", *authJwt)
+
+			if ap.ClientID != "" {
+				body.Add("client_id", ap.ClientID)
+			}
+		}
+	}
+
+	if len(ap.Scopes) > 0 {
+		body.Add("scope", strings.Join(ap.Scopes, " "))
 	}
 
 	r, err := http.NewRequest("POST", ap.TokenURL, strings.NewReader(body.Encode()))
 	if err != nil {
 		return nil, err
 	}
 	r.Header.Set("Content-Type", "application/x-www-form-urlencoded")
-	r.SetBasicAuth(ap.ClientID, ap.ClientSecret)
+
+	if ap.GrantType == "client_credentials" && ap.SigningKeyID == "" {
```

Comment:
Would it make more sense to check if ClientSecret is set than checking if SigningKey isn't?

Reply:
Indeed, fixing.